crates/sui-core/src/checkpoints/checkpoint_executor/utils.rs

```diff
@@ -244,6 +249,220 @@ pub(super) fn assert_checkpoint_not_forked(
+    FinalizeTransactions = 2,
+    ProcessCheckpointData = 3,
+    BuildDbBatch = 4,
+    CommitTransactionOutputs = 5,
+    FinalizeCheckpoint = 6,
+    UpdateRpcIndex = 7,
+    BumpHighestExecutedCheckpoint = 8,
+    End = 9,
+}
+
+impl PipelineStage {
+    pub const fn first() -> Self {
+        Self::ExecuteTransactions
+    }
+
+    fn next(self) -> Self {
+        assert!(self < Self::End);
+        // Safe to unwrap since we know the value exists and we checked it's not End
+        Self::iter().nth((self as usize) + 1).unwrap()
+    }
+
+    fn as_str(self) -> &'static str {
+        Self::VARIANTS[self as usize]
+    }
+}
+
+/// PipelineHandle is used to coordinate one iteration (sequence number) of the pipeline,
+/// starting from the first stage and ending at the last stage.
+pub(super) struct PipelineHandle {
+    seq: CheckpointSequenceNumber,
+    cur_stage: PipelineStage,
+    stages: Arc<PipelineStages>,
+    timer: Instant,
+}
+
+impl PipelineHandle {
+    fn new(stages: Arc<PipelineStages>, seq: CheckpointSequenceNumber) -> Self {
+        Self {
+            seq,
+            cur_stage: PipelineStage::first(),
+            stages,
+            timer: Instant::now(),
+        }
+    }
+
+    /// Begin at the first stage.
+    async fn begin(&mut self) {
+        assert_eq!(self.cur_stage, PipelineStage::first(), "cannot begin twice");
+        self.stages.begin(self.cur_stage, self.seq).await;
+        self.timer = Instant::now();
+    }
+
+    /// Finish a given stage and begin the next one.
+    pub async fn finish_stage(&mut self, finished: PipelineStage) {
+        let duration = self.timer.elapsed();
+        self.stages
+            .metrics
+            .stage_active_duration_ns
+            .with_label_values(&[self.cur_stage.as_str()])
+            .inc_by(duration.as_nanos() as u64);
+        assert_eq!(finished, self.cur_stage, "cannot skip stages");
+
+        self.stages.end(self.cur_stage, self.seq);
+
+        self.cur_stage = self.cur_stage.next();
+        if self.cur_stage != PipelineStage::End {
+            self.stages.begin(self.cur_stage, self.seq).await;
+        }
+    }
+
+    /// Skip to a given stage.
+    pub async fn skip_to(&mut self, stage: PipelineStage) {
+        assert!(self.cur_stage < stage);
+        while self.cur_stage < stage {
+            self.finish_stage(self.cur_stage).await;
+        }
+    }
+}
+
+/// Catch errors in which the PipelineHandle is dropped early.
+impl Drop for PipelineHandle {
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
+            assert_eq!(
+                self.cur_stage,
+                PipelineStage::End,
+                "PipelineHandle dropped without reaching end of pipeline"
+            );
+        }
+    }
+}
+
+/// A collection of watches for each stage. These are the synchronization points
+/// for the pipeline.
+pub(super) struct PipelineStages {
+    stages: [SequenceWatch; PipelineStage::End as usize],
```

Comment:
possibly not worth mentioning, but since you're already using strum, would it make sense to use strum::EnumCount here?

Reply:
hmm, but then it would be `COUNT - 1` which seems possibly more confusing?